File: src/2020/rust/src/day_02.rs
use crate::utils;

fn star1() -> u32 {
    let lines = utils::file_to_lines("../inputs/day2.txt");
    let mut result = 0;
    for line in lines {
        let parts: Vec<&str> = line.split(' ').collect();
        let bounds: Vec<u32> = parts[0]
            .split('-')
            .map(|x| x.parse::<u32>().unwrap())
            .collect();
        let letter = parts[1].chars().next().unwrap();
        let password = parts[2];

        let count = password.matches(letter).count() as u32;

        if count >= bounds[0] && count <= bounds[1] {
            result += 1;
        }
    }

    return result;
}

fn star2() -> u32 {
    let lines: Vec<String> = utils::file_to_lines("../inputs/day2.txt");
    let mut result = 0;
    for line in lines {
        let parts: Vec<&str> = line.split(' ').collect();
        let positions: Vec<u32> = parts[0]
            .split('-')
            .map(|x| x.parse::<u32>().unwrap())
            .collect();
        let letter = parts[1].chars().next().unwrap();
        let password = parts[2];

        if (password.chars().nth((positions[0] - 1) as usize).unwrap() == letter)
            ^ (password.chars().nth((positions[1] - 1) as usize).unwrap() == letter)
        {
            result += 1;
        }
    }

    return result;
}

pub fn main() {
    println!("DAY 2\n=====\nSTAR 1: {}\nSTAR 2: {}\n", star1(), star2());
}

File: src/2020/python/src/solutions/day02.py
def _star1():
    result = 0

    for line in open("../../inputs/day2.txt", "r"):
        ct, ch, pswd = line.split(" ")
        ct = ct.split("-")
        ch = ch[0]
        count = pswd.count(ch)
        if int(ct[0]) <= count <= int(ct[1]):
            result += 1

    return result


def _star2():
    result = 0

    for line in open("../../inputs/day2.txt", "r"):
        ct, ch, pswd = line.split(" ")
        ct = ct.split("-")
        ch = ch[0]
        if (pswd[int(ct[0]) - 1] == ch) ^ (pswd[int(ct[1]) - 1] == ch):
            result += 1

    return result


def day02():
    print("DAY 2\n=====\nSTAR 1: {}\nSTAR 2: {}\n".format(_star1(), _star2()))
